### PR TITLE
BLOOM Flax

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

An attempt of adding Flax implementation of BLOOM - original PR from @haileyschoelkopf #17761 

## TODOs:

- [x] alibi shifting for batched generation
- [ ] change mask fill value cc @patil-suraj 
- [x] optimize code (alibi creation + mask creation)

### CODE REVIEW

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patrickvonplaten) [inline src/transformers/models/bloom/modeling_flax_bloom.py:393]: @younesbelkada let's try to keep it simple - no need for all this manual reshaping. Our existing Flax models are quite optimized and we should follow the pattern that is already used in other models. So instead of copying from BLOOM PyTorch we should copy from OPT Flax

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patrickvonplaten) [inline src/transformers/models/bloom/modeling_flax_bloom.py]: this was a bug -> `jnp.repeat` interleaves instead of repeating for the batch

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Resolved in https://github.com/huggingface/transformers/pull/18022/commits/3444985451749540bd862910ebe7d01da8413771

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:56]: If the purpose of the `nn.relu` is to perform a `max(0, x)` operation we can replace by `jax.lax.max`

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patrickvonplaten) [inline src/transformers/models/bloom/modeling_flax_bloom.py:150]: @younesbelkada - that's the trick we usually do for left padding :-)

**Comment by patrickvonplaten** [comment]
Is this ready to merge @younesbelkada ? @sanchit-gandhi @patil-suraj could you take a final look?

**Comment by younesbelkada** [comment]
Yeah ready to merge for me, for partitioning compatibility I think that we can adress another PR for that

**Comment by younesbelkada** [comment]
Also we need to add the credits to @haileyschoelkopf for starting the contribution

File: tests/models/bloom/test_modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline tests/models/bloom/test_modeling_flax_bloom.py:42]: Head mask not used
```suggestion
def prepare_bloom_inputs_dict(config, input_ids, attention_mask=None):
```

File: tests/models/bloom/test_modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline tests/models/bloom/test_modeling_flax_bloom.py:121]: This `attention_mask` gets overridden by the one on L132 (also in [`test_flax_opt`](https://github.com/huggingface/transformers/blob/2b81f72be9fa6d69734ae27cfcbfd72b04988fe4/tests/models/opt/test_modeling_flax_opt.py#L129), from which this was copied). Can remove this line.

**Review by sanchit-gandhi — COMMENTED** [review]
@younesbelkada some comments regarding the tests!

File: tests/models/bloom/test_modeling_flax_bloom.py
Comment (younesbelkada) [reply tests/models/bloom/test_modeling_flax_bloom.py]: Thanks for the catch!

Comment (younesbelkada) [reply tests/models/bloom/test_modeling_flax_bloom.py]: Also changed the function name in a test

**Comment by sanchit-gandhi** [comment]
Modelling file and tests LGTM. Am in agreement that the the named axes should come as a follow-up PR - better to modify a script that's already in-place and working rather than merging a complete overhaul of how we structure our code in one go.

File: tests/models/bloom/test_modeling_flax_bloom.py
Comment (younesbelkada) [reply tests/models/bloom/test_modeling_flax_bloom.py:121]: Just removed the L132 and seems to work fine

Comment (sanchit-gandhi) [reply tests/models/bloom/test_modeling_flax_bloom.py:121]: The one on L132 was superseded by a second definition of `attention_mask` 3/4 lines later, and so was un-used.

Comment (younesbelkada) [reply tests/models/bloom/test_modeling_flax_bloom.py:121]: yeah so I removed the one on the L124 and directly the one from  `inputs_dict`

**Comment by patrickvonplaten** [comment]
Good to merge for me once @patil-suraj and @sanchit-gandhi give the green light here :-)

File: tests/models/bloom/test_modeling_flax_bloom.py
Comment (younesbelkada) [reply tests/models/bloom/test_modeling_flax_bloom.py]: Sorry removed the wrong line actually, should be fixed in cecfcb158e7bfc2558d6f9f1ea7f270468e0dbc8

**Review by sanchit-gandhi — APPROVED** [review]
LGTM, thanks everyone! And special mention to @haileyschoelkopf who kicked-off this PR 🤗

For reference, the modelling code used in the deployment of Flax BLOOM is currently in a standalone repository: https://github.com/huggingface/bloom-jax-inference/blob/main/bloom_inference/modeling_bloom/modeling_bloom.py
It employs 2D model, activation and data sharding, and a faster generation method with more efficient indexing. These changes will be gradually brought across to the Transformers modelling code with the introduction of named axes support.

**Comment by younesbelkada** [comment]
Thanks all for your comments! 
I had to fix a small test since the main modeling code has changed a bit in #18344 !
I think this PR is ready to be merged 💪 cc @patil-suraj

**Comment by patrickvonplaten** [comment]
@patil-suraj @sanchit-gandhi - should we merge this one?

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_pytorch_utils.py:112]: sounds good to me!

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patrickvonplaten) [inline src/transformers/models/bloom/modeling_flax_bloom.py:538]: @younesbelkada - this should be rewritten in with `def setup(...)` function as stated in the comment. Could you take care of it?

**Review by patrickvonplaten — COMMENTED** [review]
One small change regarding `setup` would be nice here as stated in the comment

**Comment by sanchit-gandhi** [comment]
Good for me! Awaiting the green light from @patil-suraj who requested a look over the modelling code prior to merging. Regarding `setup`, I previously asked in the G Chat with the JAX/Flax guys how to make this work with `scan`. Think it went under the radar, so will bring this issue back up!

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:275]: Is this still standing ?
Here we could simply upcast the attention weights to `fp32` before doing softmax to achieve this.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:315]: These two are not used
```suggestion
```

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:349]: Is this option used ? When I was working on it all the model had the same value for `apply_residual_connection_post_layernorm`. If that's the case maybe we could remove this.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:224]: `layer_number` is not used anywhere.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:383]: `layer_number` is not used anywhere.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:537]: should be taken care of

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:615]: We shouldn't pass the `dtype` to embeds, this will keep the embed weights in that `dtype`,  but it's fine if that's what we want here.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (patil-suraj) [inline src/transformers/models/bloom/modeling_flax_bloom.py:637]: `input_ids` are already casted to `i4` in the pre-trained model class no ?

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for adding BLOOM in Flax !
LGTM, but some todo's are standing, should be good to merge once those are addressed.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: I should have proposed a fix in dda4a08e85680897dce646b5669603e7e7246c6e
But not sure if this breaks anything related to `use_scan`, will check later if there is a test for that

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: I think yes, since it is still in place in the original modeling code. 
Proposed a simple fix in: c685442e3548dcf6c9ebaeb91a17a480eb683ba5

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Fixed in 1999df780452ff4b0830c1d3eed49f098c083076 ! Could not managed to commit directly from your proposed change since the diff was outdated

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py:352]: I think that we should keep it in case people train bloom models with different values for `apply_residual_connection_post_layernorm` so I would keep it I think for now

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Should be fixed in fc371268add3de5cb68ad6ecb44cbbf6d9dde90a

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py:551]: Let's move the discussion at https://github.com/huggingface/transformers/pull/18022#discussion_r954115059

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py:643]: Some tests where failing when removing the casting, probably the `input_ids` are not casted to `i4` in the pre-trained model class - will investigate that

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Managed to fix the test! 
I had to refactor a bit the code and use the following versions of `jax` and `flax`
```
jax.__version__
>>> ‘0.3.17’
flax.__version__
>>> ‘0.6.0’
```

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Thanks for pointing out this issue! Finally fixed in 05c05212fb629487227b5e20b16a57995b3709c7

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py:622]: I think that this is what we want here 💪

**Comment by younesbelkada** [comment]
Hi @sanchit-gandhi @patil-suraj @patrickvonplaten 
I think that the PR is ready for a final round of review! 💪  I proposed a potential solution to re-write the model definition with `setup` but I am unsure how this fits the HF+Jax approach.

**Comment by patrickvonplaten** [comment]
@sanchit-gandhi @patil-suraj could you check here?

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [reply src/transformers/models/bloom/modeling_flax_bloom.py:622]: We should pass the `dtype` arg to keep the weights in bfloat16 for half precision inference.

Also related: https://github.com/huggingface/transformers/pull/18462

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:543]: Technically we don't actually need this `if` statement: `self.scan_fn` will only be instantiated if used in the `__call__` method.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:552]: We can remove this TODO
```suggestion
```

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:359]: This arg defaults to `None` everywhere and is not used in any of the `__call__` methods - can we remove it @younesbelkada?

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:576]: See here how `layer_past` is always set to `None` - if we trace it back to `FlaxBloomBlock` we see that it's not used anywhere in the BLOOM model.

**Review by sanchit-gandhi — COMMENTED** [review]
Almost there! Just some small refactoring suggestions to clean the code up a bit

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Thanks! Should be fixed in b002f6494271dbaae5ee01ce0414c9e8d76435a1

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Yes we can! Fixed in 0b70bf22ee732fb741c2b2184164227d6b589470

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (sanchit-gandhi) [reply src/transformers/modeling_flax_pytorch_utils.py:117]: Is this necessary? Looks to me to be redundant! We check for `torch` import in `load_pytorch_checkpoint_in_flax_state_dict`:
https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/modeling_flax_pytorch_utils.py#L45-L53
And then call `convert_pytorch_state_dict_to_flax` 9 lines later:
https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/modeling_flax_pytorch_utils.py#L62
So this check in `convert_pytorch_state_dict_to_flax`is superseded by the parent function `load_pytorch_checkpoint_in_flax_state_dict`.

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:113]: This is currently quite difficult to understand with the triple nested function structure! Could we:
1) Add a docstring similar to [modeling_bloom.py#L89](https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/models/bloom/modeling_bloom.py#L89)
2) Unravel the triple nested function (copy the relevant logic from [modeling_bloom.py#L104-L118](https://github.com/huggingface/transformers/blob/d0d5aee1dd26de108a901921b9df19b889430645/src/transformers/models/bloom/modeling_bloom.py#L104-L118))

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:563]: We can uncomment these checks

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:561]: We've currently left `scan` in the modelling code. Part of me thinks we should remove it for Transformers for the following reasons: 
1. `scan` adds a lot of boilerplate code that isn't very easy to understand
2. It is only beneficial for compile times when the model size is large, mostly when training and less so for inference

In the latter case, users will also likely shard the model, meaning they can employ the standalone code in [bloom-jax-inference](https://github.com/huggingface/bloom-jax-inference/blob/main/bloom_inference/modeling_bloom/modeling_bloom.py) where we can retain`scan`.

We also found generation time to be slower when using `scan` vs not using it (despite a faster compile time). The generation time will amortise the compile time in any use case of Flax BLOOM.

Given that the philosophy of Transformers is functional, easy-to-understand code that is not necessarily fully optimised, I'm in favour of stripping `scan` from Flax BLOOM and leaving it to [bloom-jax-inference](https://github.com/huggingface/bloom-jax-inference/blob/main/bloom_inference/modeling_bloom/modeling_bloom.py) to serve users that want to deploy larger variants of the model.

WDYT @patrickvonplaten @patil-suraj @younesbelkada?

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:609]: Fine to use this value directly in the `nn.Embed` rather than separately defining a variable! No harm in what you've done though :)

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:285]: Can omit this arg as we've set it to its default value (`None`)

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:642]: Don't see these variables being used

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for addressing the previous comments @younesbelkada. Two things from this round of review:
1. Could we tidy up the `build_alibi_tensor_flax` function to avoid a triple nested function? Much of the logic can be copied over from PyTorch BLOOM!
2. Big question for me is whether we keep `scan` or not - I'm in favour of removing it for Flax BLOOM (see comments below)

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (younesbelkada) [reply src/transformers/modeling_flax_pytorch_utils.py:117]: Yeah let's remove it - I just had to add `# noqa: F821` to make the quality test pass here

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Done in 38c59465f44e52546b7b88c29aaf76429f87e94f

File: src/transformers/models/bloom/modeling_flax_bloom.py
Comment (younesbelkada) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Proposed a major refactor in 38c59465f44e52546b7b88c29aaf76429f87e94f !

**Review by younesbelkada — COMMENTED** [review]
Should have addressed your new suggestions @sanchit-gandhi ! Here I mainly focused on refactoring a bit the `build_alibi` function to match the implementational style of `Pytorch` -> this way it seems to be more readable !
I will leave you and @patil-suraj and @patrickvonplaten decide regarding the scan feature and happy to remove it once we agree on that!
Can also confirm the slow tests/conversion tests pass ;)

**Comment by younesbelkada** [comment]
Should we merge this one ? cc @patrickvonplaten @patil-suraj @sanchit-gandhi

**Comment by wangkuiyi** [comment]
Both https://github.com/huggingface/transformers/pull/17761 and this PR show a lot of work. Why isn't it merged?

**Comment by sanchit-gandhi** [comment]
Adding this to my TODOs

**Pull request closed without merging**
